omnigibson/transition_rules.py

```diff
@@ -1210,35 +1252,34 @@ def _execute_recipe(cls, container, recipe, in_volume):
         Returns:
             TransitionResults: Results of the executed recipe transition
         """
-        t_results = TransitionResults()
+        objs_to_add, objs_to_remove = [], []
 
         # Compute total volume of all contained items
         volume = 0
 
         # Remove all recipe system particles contained in the container
-        for system in PhysicalParticleSystem.get_active_systems():
+        contained_particles_state = container.states[ContainedParticles]
+        for system in PhysicalParticleSystem.get_active_systems().values():
             if container.states[Contains].get_value(system):
-                volume += contained_particles_state.get_value()[0] * np.pi * (system.particle_radius ** 3) * 4 / 3
-                container.states[Filled].set_value(system, False)
-        for system in VisualParticleSystem.get_active_systems():
+                volume += contained_particles_state.get_value(system).n_in_volume * np.pi * (system.particle_radius ** 3) * 4 / 3
+                container.states[Contains].set_value(system, False)
+        for system in VisualParticleSystem.get_active_systems().values():
             group_name = system.get_group_name(container)
             if group_name in system.groups and system.num_group_particles(group_name) > 0:
                 system.remove_all_group_particles(group=group_name)
 
         # Remove either all objects or only the recipe-relevant objects inside the container
-        t_results.remove = np.concatenate([
-                cls._OBJECTS[np.where(in_volume[cls._CATEGORY_IDXS[obj_category]])[0]]
-                for obj_category in recipe["input_objects"].keys()
-            ]).tolist() if cls.ignore_nonrecipe_objects else cls._OBJECTS[np.where(in_volume)[0]].tolist()
+        objs_to_remove.extend(np.concatenate([
+            cls._OBJECTS[np.where(in_volume[cls._CATEGORY_IDXS[obj_category]])[0]]
+            for obj_category in recipe["input_objects"].keys()
+        ]) if cls.ignore_nonrecipe_objects else cls._OBJECTS[np.where(in_volume)[0]])
         volume += sum(obj.volume for obj in objs_to_remove)
 
         # Define callback for spawning new objects inside container
         def _spawn_object_in_container(obj):
-            # We will either sample Inside or OnTop, based on the relative AABB extents of the container link
-            # vs. the object link
-            container_aabb = container.states[ContainedParticles].link.aabb_extent
-            obj_aabb = obj.aabb_extent
-            state = Inside if np.all(container_aabb > obj_aabb) else OnTop
+            # For simplicity sake, sample only OnTop
+            # TODO: Can we sample inside intelligently?
+            state = OnTop
```

Comment:
hmm this is very tricky. Does it mean if I bake a pie, the pie will be sampled on top of the oven? Maybe this should be included as part of the annotation for each transition rule (i.e. where do you want the output product to be sampled w.r.t the container)?